daiquiri/query/viewsets.py

```diff
@@ -229,34 +231,50 @@ def columns(self, request, pk=None, format_key=None):
 
         return Response(job.columns())
 
-    @action(detail=True, methods=['get'], url_path='download/(?P<download_id>[A-Za-z0-9\-]+)', url_name='download')
-    def download(self, request, pk=None, download_id=None):
+    @action(detail=True, methods=['get'], url_name='download',
+            url_path=r'download/(?P<download_key>[a-z\-]+)/(?P<download_job_id>[A-Za-z0-9\-]+)')
+    def download(self, request, pk=None, download_key=None, download_job_id=None):
+        print(pk, download_key)
+
         try:
-            job = self.get_queryset().get(pk=pk)
+            self.get_queryset().get(pk=pk)
         except QueryJob.DoesNotExist:
             raise NotFound
 
+        download_config = get_download_config(download_key)
+        if download_config is None:
+            raise ValidationError({'download': "Not supported."})
+
+        download_job_model = import_class(download_config['model'])
+
         try:
-            download_job = job.downloads.get(pk=download_id)
-        except DownloadJob.DoesNotExist:
+            download_job = download_job_model.objects.get(query_job=pk, pk=download_job_id)
+        except download_job_model.DoesNotExist:
             raise NotFound
 
```

Comment:
After some reflection I could figure out what this piece of code does. I find it a bit wordy, and misleading what about something like:

```python
download_job_model = get_download_job_model(download_key)
```

with 

```python
def get_download_job_model(key):
    
    try:
        # this is a bit complex for just a selection of a dictionary item, but that is fine for me.
        download_config = next(filter(lambda c: c['key'] == download_key, settings.QUERY_DOWNLOADS))
    except StopIteration:
        ValidationError({'download type': "Not supported."}) # maybe specifying the key we got (in case of typos)

        return import_class(download_config['model'])
```

It could get even clearer but I think this could be a good balance.

Reply:
The problem is that I need `download_config.get('params', [])` in the second case, and I try to raise HTTP exceptions in the view to keep the utils functions simple. I could add the import_class part to the utils function and then use `download_config['model']` (and rename the current `model` setting to `model_class`), but this is not much easiear. I updated the Error Message though.

Reply:
I see the argument of keeping `utils.py` simple as well as raising HTTP within `views.py`
fine for me.